lib/TPP/ConvertPackOptimization.cpp

```diff
@@ -0,0 +1,202 @@
+/// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+////
+////===----------------------------------------------------------------------===//
+//
+//
+
+#include "TPP/BuilderUtils.h"
+#include "TPP/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+using namespace mlir::tpp;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+
+namespace {
+
+struct MatchResult {
+  bool haveMatch;
+  int index;
+};
+
+/*
+ * Returns a list of indices that match the inner dimensions' position
+ */
+SmallVector<MatchResult> innerDimMatchesIndex(tensor::PackOp packOp) {
+  SmallVector<MatchResult> result;
+  int innerTiles = 0;
+  for (size_t i = 0; i < packOp.getSourceType().getShape().size(); i++) {
+    if (packOp.getInnerDimsPos().size() > 0) {
+      bool haveMatch = false;
+      for (size_t j = 0; j < packOp.getInnerDimsPos().size(); j++) {
+        if (i == (size_t)packOp.getInnerDimsPos()[j]) {
+          MatchResult tempResult;
+          tempResult.haveMatch = true;
+          haveMatch = true;
+          tempResult.index = innerTiles++;
+          result.push_back(tempResult);
+          break;
+        }
+      }
+      if (!haveMatch) {
+        MatchResult tempResult;
+        tempResult.haveMatch = false;
+        tempResult.index = -1;
+        result.push_back(tempResult);
+      }
+    }
+  }
+  return result;
+}
+
+struct ConvertPackOptimizationOp : public OpRewritePattern<tensor::PackOp> {
+  using OpRewritePattern<tensor::PackOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(tensor::PackOp packOp,
+                                PatternRewriter &rewriter) const override {
+    if (packOp.getStaticInnerTiles().size() <= 0)
+      return failure();
+    auto shape = packOp.getSourceType().getShape();
+    int numLoops = shape.size();
+
+    SmallVector<MatchResult> haveMatched = innerDimMatchesIndex(packOp);
+    for (int i = 0; i < numLoops; i++) {
+      if (haveMatched[i].haveMatch) {
+        if (shape[i] < packOp.getStaticInnerTiles()[haveMatched[i].index]) {
+          return failure();
+        }
+      }
+    }
+
+    auto zero = getConstIndex(rewriter, 0);
+    auto one = getConstIndex(rewriter, 1);
+
+    SmallVector<Value> lbs(numLoops, zero);
+    SmallVector<Value> ubs;
+    SmallVector<Value> steps(numLoops, one);
+
+    for (int i = 0; i < numLoops; i++) {
+      if (haveMatched[i].haveMatch && shape[i] != ShapedType::kDynamic &&
+          packOp.getStaticInnerTiles()[haveMatched[i].index] !=
+              ShapedType::kDynamic) {
+        ubs.push_back(getConstIndex(
+            rewriter,
+            shape[i] / packOp.getStaticInnerTiles()[haveMatched[i].index]));
+
+      } else {
+        ubs.push_back(getConstIndex(rewriter, shape[i]));
+      }
+    }
+
+    SmallVector<Value> reduc = {
```

Comment:
You don't need this `reduc`. I think you can pass directly `packOp.getDest()`.

Reply:
removed.